### PR TITLE
Exposing chinese_whispers directly, closes #1642

### LINKED ISSUE #1642: Expose chinese_whispers to Python
Currently, there is http://dlib.net/python/index.html#dlib.chinese_whispers_clustering exposed which takes descriptors and do clustering. If user wants to give it 128D vectors itself (or any other generic edges, for that matter), there is no easy way, as you cannot convert to C++ descriptors from Python. Does it makes sense to either:
* expose `dlib::chinese_whispers` directly from Python API, as `dlib.chinese_whispers(edges: list) -> label list`
* expose `dlib::chinese_whispers` (or overload ``dlib::chinese_whispers_clustering`) to accept list of lists (all elements should have same size), not just list of C++ objects. And C++ would calculate euclidean distance to get edges

Does this sound something that can be accepted in dlib? And with what level of abstraction?

### PR DESCRIPTION
**Pull request by stalker314314**
Boilerplate code to expose chinese_whispers from Python

### CODE REVIEW

File: tools/python/src/face_recognition.cpp [hunk 1]
```diff
@@ -206,6 +206,35 @@ py::list chinese_whispers_clustering(py::list descriptors, float threshold)
...
+        py::tuple t = edges[idx].cast<py::tuple>();
+        if (len(t) != 2)
+        {
+            PyErr_SetString( PyExc_IndexError, "Input must be a list of tuples with exactly 2 elements.");
+            throw py::error_already_set();
+        }
+        size_t i = t[0].cast<size_t>();
+        size_t j = t[1].cast<size_t>();
+
+        edges_pairs.push_back(sample_pair(i,j));
```
[the remark this answers is not in the document]
Comment (stalker314314) [reply tools/python/src/face_recognition.cpp]: Done. Did it in same method (e.g. didn't added `chinese_whispers_with_distance` flavor). Check it out

File: tools/python/src/face_recognition.cpp [hunk 2]
```diff
@@ -296,5 +325,8 @@ void bind_face_recognition(py::module &m)
     m.def("chinese_whispers_clustering", &chinese_whispers_clustering, py::arg("descriptors"), py::arg("threshold"),
         "Takes a list of descriptors and returns a list that contains a label for each descriptor. Clustering is done using dlib::chinese_whispers."
         );
+    m.def("chinese_whispers", &chinese_whispers_raw, py::arg("edges"),
+        "Takes a list of edges (edges are pairs of connected vertices presented as tuples) and returns a list that contains a label for each vertex. Offers direct access to dlib::chinese_whispers."
```
[the remark this answers is not in the document]
Comment (stalker314314) [reply tools/python/src/face_recognition.cpp]: OK, had another stab. Did my best:) Still, trying to explain something like this and trying to be as laconic as possible is not easy:D If you have better wording/idea, please, let me know

**Comment by stalker314314** [comment]
> Also please add some unit tests

Tests added (of course, just a basic ones; I guess testing `chinese_whispers` algo itself was not your initial intention)

**Comment by davisking** [comment]
Sweet, thanks for updating the PR.  The changes look good.  I think it just needs one more test that checks the output for know correct results.  Like check if the two vertices get the same labels when connected and different ones when disconnected.

**Comment by stalker314314** [comment]
> Like check if the two vertices get the same labels when connected and different ones when disconnected.

Removed previous (more) trivial check that args are used and now doing full checks for results. I hope this is what you had in mind?

**Comment by stalker314314** [comment]
Friendly ping:)

**Comment by davisking** [comment]
Nice, looks good.  Sorry for the delay, was pretty busy the last few days :)

**Pull request merged**